File: java/cours/src/session2/td/agregation/NaturelInductif.java
package session2.td.agregation;

public interface NaturelInductif extends EtatNaturelPur {
	
	@Override
	default EtatNaturelPur creerZero() {
		return new ZeroInductif();
	}
	
	@Override
	default EtatNaturelPur creerSuccesseur(EtatNaturelPur pred) {
		return new SuccInductif(pred);
	}
	
	@Override
	default EtatNaturelPur creerNatAvecValeur(int val) {
		if (val == 0)
			return this.creerZero();
		return this.creerSuccesseur(this.creerNatAvecValeur(val - 1));
	}
	
	
}

File: typescript/session2/agregation/NaturelInductif.ts
import {EtatNaturelPur} from "./EtatNaturelIPur";
import {ZeroInductif} from "./ZeroInductif";
import {SuccInductif} from "./SuccInductif";

export abstract class NaturelInductif extends EtatNaturelPur {

    creerZero() : EtatNaturelPur {
    return new ZeroInductif();
}

creerSuccesseur(pred : EtatNaturelPur) : EtatNaturelPur {
    return new SuccInductif(pred);
}

creerNatAvecValeur(val : number) : EtatNaturelPur{
    if (val == 0)
        return this.creerZero();
    return this.creerSuccesseur(this.creerNatAvecValeur(val - 1));
}


}
